[NVIDIA GPU] Add Bitcast to collective pipeliner acceptable users

terryysun opened the pull request:
Right now collective pipeliner will filter user type when determining if a value can be pushed to the next iteration of the loop. Bitcast is not in the acceptable users, which causes the collective pipeliner not to kick in under some situations. This PR adds Bitcast to acceptable users.

golechwierowicz commented:
Hey,

on which model do you see a lack of this support leading to collective pipeliner not kicking in?

In general I think it is a safe change, but please add a test.

terryysun commented:
> Hey,
> 
> on which model do you see a lack of this support leading to collective pipeliner not kicking in?
> 
> In general I think it is a safe change, but please add a test.

It's an internal implementation of MaxText, the absence of Bitcast caused collective pipeliner not to kick in in backward pass. Added a test for a situation where this change is useful.

frgossen commented on xla/service/collective_pipeliner_test.cc:
```diff
@@ -181,6 +181,63 @@ ENTRY entry {
+}
+
+while_body {
+  param = (s32[], bf16[3,8,128], bf16[3,8,128]) parameter(0)
```
Can you simplify this test case and give meaningful names to the values? 
It looks to me that there are a lot of ops in this that are not related to pipelining and are not relevant to the test case. 
I think a comment could also help understand what is tested here.

frgossen reviewed (changes requested):
Thank you for adding the test case :)  
I think we can simplify it a bit still

golechwierowicz commented:
frgossen@ PTL

frgossen commented:
> @frgossen PTL

I don't think the last comments were addressed yet

terryysun commented:
> > @frgossen PTL
> 
> I don't think the last comments were addressed yet

I've simplified the test HLO with some comments and renamed the variables, @frgossen could you comments on things you find unaddressed? Thanks!

frgossen commented:
Sorry, I was expecting some reply to the comments, even just a simple `Done`. Based on the commits alone, I find it hard to tell when the PR is ready for the next round of review. 

Thanks for simplifying the test case. Will take another look!

frgossen commented on xla/service/collective_pipeliner_test.cc:
```diff
@@ -181,6 +181,59 @@ ENTRY entry {
+
+  all-reduce.0 = bf16[1,8,128] all-reduce(sliced-parameter.1), replica_groups={}, to_apply=add, channel_id=1
+  bitcasted-all-reduce.0 = u16[3,8,128] bitcast(all-reduce.0)
+  all-reduce.1 = bf16[3,8,128] bitcast(bitcasted-all-reduce.0)
```
This is the one that depends on a bitcast and you want to see it pipelined, right? 
To test your change, I think you need no more than these two ops in the loop, do you?

terryysun replied:
yes this is the part we want to test. but in order to trigger collective pipeliner, we need the pattern of `compute op` (this is what to overlap, dynamic-slice here) `-> collective op` (the other key part to overlap, all-reduce here) `-> elementwise formatting op` (what we are updating, bitcast here) `-> dynamic-update-slice` (because collective pipeliner will reuse the output buffer for intermediate result), I believe it's already the minimal set of ops for this pattern.

frgossen commented on xla/service/collective_pipeliner_test.cc:
```diff
@@ -181,6 +181,59 @@ ENTRY entry {
+while_body {
+  param = (s32[], bf16[3,8,128], bf16[3,8,128]) parameter(0)
+  current-loop-index = s32[] get-tuple-element(param), index=0
+  parameter.0 = bf16[3,8,128] get-tuple-element(param), index=1
```
If these parameters are relevant to the test, can you give them a name that explains their role?

terryysun replied:
renamed

frgossen commented on xla/service/collective_pipeliner_test.cc:
```diff
@@ -181,6 +181,59 @@ ENTRY entry {
+  EXPECT_TRUE(RunOptimizer(module.get(), /*last_run=*/true).value());
+  XLA_VLOG_LINES(1, module->ToString());
+  const HloInstruction* root = module->entry_computation()->root_instruction();
+  EXPECT_THAT(root, op::DynamicUpdateSlice(_, _, _, _, _));
```
Can we check a bit more here? I think the test case should check the `all-reduce`s that you want to see pipelined.

terryysun replied:
added

frgossen reviewed (changes requested):
Thanks for simplifying the test. 
I think can still be simplified quite a bit more to test your change.

terryysun commented:
@frgossen could you take another look? Thanks!
